### PR TITLE
[V1] Use FlashInfer Sampling Kernel for Top-P & Top-K Sampling

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR brings FlashInfer's top-p & top-k sampling kernel to the V1 sampler.

The PyTorch-native implementation of top-p and top-k sampling requires sorting the entire logits tensor, which could be very large when the batch size is large. As a result, the sorting op can take ~20% performance overhead in high-throughput settings. By leveraging the FlashInfer's kernel, we can reduce this down to almost zero.

ShareGPT benchmark: 
* `main` w/o top-p & top-k: 47.40 reqs/s
* `main` w/ top-p: 40.20 reqs/s
* This RP w/ top-p: 47.15 reqs/s

### CODE REVIEW

File: tests/v1/sample/test_sampler.py
Comment (WoosukKwon) [inline tests/v1/sample/test_sampler.py:1]: I modified the tests here because the sampler returns the original logprobs instead of the final log probs after applying penalties and scaling.

**Comment by WoosukKwon** [comment]
@sroy745 @comaniac Could you please take a look?

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        super().__init__()
+        if current_platform.is_cuda:
+            if use_flashinfer:
+                self.forward = self.forward_cuda
+            else:
+                logger.warning(
+                    "FlashInfer is not available. Falling back to the PyTorch-"
+                    "native implementation of top-p & top-k sampling. For the "
+                    "best performance, please install FalshInfer.")
+                self.forward = self.forward_native
```
Comment (comaniac) [inline vllm/v1/sample/ops/topk_topp_sampler.py:30]: IIRC previously the FlashInfer sampler outputs different results as the native one, so we make it configurable even FlashInfer is installed, but I'm not sure if this is still the case. cc @yzh119.

File: vllm/v1/sample/ops/topk_topp_sampler.py
Comment (comaniac) [inline vllm/v1/sample/ops/topk_topp_sampler.py:1]: Could you add docstring to the functions in this file?

File: vllm/v1/sample/sampler.py [hunk 2]
```diff
@@ -1,43 +1,50 @@
...
-            _apply_penalties(logits, sampling_metadata.prompt_token_ids,
-                             sampling_metadata.presence_penalties,
-                             sampling_metadata.frequency_penalties,
-                             sampling_metadata.repetition_penalties,
-                             sampling_metadata.output_token_ids)
+        needs_logprobs = sampling_metadata.max_num_logprobs > 0
+        if needs_logprobs:
+            # NOTE: We need to clone the tensor because the below ops may
+            # modify the logits tensor in-place.
+            orig_logits = logits.clone()
```
Comment (comaniac) [inline vllm/v1/sample/sampler.py:29]: Since logprobs is a separated path, may be we could move the next `needs_logprobs` here so that `orig_logits` can be released earlier to save some memory?

```python
if needs_logprobs
    logprobs = self.get_logprobs(logits.clone())
    ...
else:
    logprobs = None
    ...
```

**Review by comaniac — APPROVED** [review]
Overall LGTM

File: vllm/v1/sample/ops/topk_topp_sampler.py
Comment (WoosukKwon) [reply vllm/v1/sample/ops/topk_topp_sampler.py:1]: Good point. Added docstrings & comments.

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/ops/topk_topp_sampler.py]: Yeah we actually have `VLLM_USE_FLASHINFER_SAMPLER` env variable which is False by default.

I do want users to benefit from this by default though, given the considerable perf difference. Can we enable this env variable by default? Otherwise, I would personally prefer to leave this as is (v1 using it whenever available).

File: vllm/v1/sample/sampler.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/sample/sampler.py]: Good idea! Updated the code.

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 1 above)
Comment (comaniac) [reply vllm/v1/sample/ops/topk_topp_sampler.py]: Hmm can we set default to True for v1 only?

Comment (WoosukKwon) [reply vllm/v1/sample/ops/topk_topp_sampler.py]: @comaniac Good point. I made VLLM_USE_FLASHINFER_SAMPLER None by default and interpret it as True only for V1.

Comment (comaniac) [reply vllm/v1/sample/ops/topk_topp_sampler.py]: Thanks! Looks all good to me

**Comment by sirouk** [comment]
@WoosukKwon thanks for this 🙏 

Do you think this could be related?
https://github.com/vllm-project/vllm/issues/18811
